builtin/providers/aws/resource_customer_gateway.go

```diff
@@ -0,0 +1,126 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/mitchellh/goamz/ec2"
+)
+
+func resourceAwsCustomerGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCustomerGatewayCreate,
+		Read:   resourceAwsCustomerGatewayRead,
+		Update: resourceAwsCustomerGatewayUpdate,
+		Delete: resourceAwsCustomerGatewayDelete,
+
+		Schema: map[string]*schema.Schema{
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"ip_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"bgp_asn": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default: 65000,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCustomerGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	ec2conn := meta.(*AWSClient).ec2conn
+
+	createOpts := &ec2.CreateCustomerGateway{
+		Type:            d.Get("type").(string),
+		IpAddress:       d.Get("ip_address").(string),
+		BgpAsn:          d.Get("bgp_asn").(int),
+	}
+
+	// Create the gateway
+	log.Printf("[DEBUG] Creating customer gateway")
+	resp, err := ec2conn.CreateCustomerGateway(createOpts)
+
+	if err != nil {
+		return fmt.Errorf("Error creating customer gateway: %s", err)
+	}
+
+	// Get the ID and store it
+	customerGateway := &resp.CustomerGateway
+	log.Printf("[INFO] CustomerGateway ID: %s", customerGateway.CustomerGatewayId)
+	d.SetId(customerGateway.CustomerGatewayId)
+
+	return resourceAwsCustomerGatewayUpdate(d, meta)
+}
+
+func resourceAwsCustomerGatewayUpdate(d *schema.ResourceData, meta interface{}) error {
+	ec2conn := meta.(*AWSClient).ec2conn
+	d.Partial(true)
```

Comment:
Since this is called from `Create` as well, I think we need to set Partial to true in create, and use `SetPartial` as needed for the various fields in create. It is safe to call `Partial(true)` multiple times.